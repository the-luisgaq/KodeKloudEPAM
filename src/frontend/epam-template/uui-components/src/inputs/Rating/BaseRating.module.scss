.container {
    position: relative;
    cursor: pointer;

    &:focus {
        outline: none;
    }

    &--readonly {
        cursor: default;
    }
}


