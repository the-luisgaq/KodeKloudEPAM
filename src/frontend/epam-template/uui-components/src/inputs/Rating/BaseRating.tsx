import * as React from 'react';
import {
    cx, IDisableable, ICanBeReadonly, IHasCX, uuiMod, IHasRawProps, IHasForwardedRef, IControlled,
} from '@epam/uui-core';
import css from './BaseRating.module.scss';

export interface BaseRatingProps<TValue>
    extends IHasCX,
    IDisableable,
    IControlled<TValue>,
    ICanBeReadonly,
    IHasRawProps<React.HTMLAttributes<HTMLDivElement>>,
    IHasForwardedRef<HTMLDivElement> {
    from?: number;
    to?: number;
    /**
     * @default 1
     */
    step?: 0.5 | 1;
    renderRating?: (rating: number, markWidth: number, numberOfMarks: number) => React.ReactNode;
}

interface BaseRatingState {
    rating?: number;
    containerWidth?: number;
}

export class BaseRating extends React.Component<BaseRatingProps<number>, BaseRatingState> {
    container: HTMLElement | null;
    isPointerMoveSubscribed = false;

    constructor(props: BaseRatingProps<number>) {
        super(props);
        this.state = {};
    }

    componentDidMount() {
        this.setState({
            rating: this.checkRating(this.props.value),
            containerWidth: this.getContainerWidth(),
        });
    }

    componentDidUpdate(prevProps: BaseRatingProps<number>) {
        if (this.state.containerWidth !== this.getContainerWidth()) {
            this.setState({
                containerWidth: this.getContainerWidth(),
            });
        }
        if (this.props.value !== prevProps.value) {
            this.setState({
                rating: this.checkRating(this.props.value),
            });
        }
    }

    getContainerWidth() {
        return this.container?.offsetWidth;
    }

    getContainerOffsetLeft() {
        return this.container.getBoundingClientRect().left;
    }

    getNumberOfMarks() {
        const step = this.props.step || 1;
        return (this.props.to - this.props.from) / step + 1;
    }

    getMarkWidth() {
        return this.state.containerWidth / this.getNumberOfMarks();
    }

    checkRating(rating: number): number {
        if (!rating && rating !== 0) {
            return rating;
        } else if (rating < this.props.from - this.props.step) {
            return this.props.from;
        } else if (rating > this.props.to) {
            return this.props.to;
        } else {
            return rating;
        }
    }

    getRatingFromWidth(width: number): number {
        const step = this.props.step || 1;

        return step * Math.ceil(width / this.getMarkWidth()) + (this.props.from - step);
    }

    onPointerMove(e: React.PointerEvent<HTMLDivElement>) {
        if (!this.isPointerMoveSubscribed) {
            window.addEventListener('pointermove', this.handleWindowPointerMove);
            this.isPointerMoveSubscribed = true;
        }

        const width = e.pageX - this.getContainerOffsetLeft();
        if (width > 0 && width <= this.state.containerWidth) {
            this.setState({ rating: this.checkRating(this.getRatingFromWidth(width)) });
        }
    }

    handleWindowPointerMove = (e: PointerEvent) => {
        if (!this.container) return;
        const { clientX, clientY } = e;
        const rect = this.container.getBoundingClientRect();
        const inside = clientX >= rect.left
            && clientX <= rect.right
            && clientY >= rect.top
            && clientY <= rect.bottom;

        if (!inside) {
            this.resetRatingToValue();
            window.removeEventListener('pointermove', this.handleWindowPointerMove);
            this.isPointerMoveSubscribed = false;
        }
    };

    unsubscribePointerMove = () => {
        if (this.isPointerMoveSubscribed) {
            window.removeEventListener('pointermove', this.handleWindowPointerMove);
            this.isPointerMoveSubscribed = false;
        }
    };

    resetRatingToValue = () => {
        this.setState({ rating: this.checkRating(this.props.value) });
    };

    onPointerUp(e: React.PointerEvent<HTMLDivElement>) {
        const width = e.pageX - this.getContainerOffsetLeft();
        this.props.onValueChange(this.checkRating(this.getRatingFromWidth(width)));
    }

    onTouchEnd(e: React.TouchEvent<HTMLDivElement>) {
        const touch = e.changedTouches[0];
        const width = touch.pageX - this.getContainerOffsetLeft();
        this.props.onValueChange(this.checkRating(this.getRatingFromWidth(width)));
    }

    onKeyDown(e: React.KeyboardEvent<HTMLDivElement>) {
        const { from, to } = this.props;
        const { rating } = this.state;
        const step = this.props.step || 1;

        if (e.key === 'ArrowLeft') {
            if (rating - step < from) return;
            else this.setState({ rating: rating - step });
        } else if (e.key === 'ArrowRight') {
            if (rating + step > to) return;
            else this.setState({ rating: rating + step });
        }
    }

    handleRef = (container: HTMLDivElement) => {
        this.container = container;
        (this.props.forwardedRef as React.RefCallback<HTMLDivElement>)?.(container);
    };

    onPointerLeave = () => {
        if (this.props.isReadonly || this.props.isDisabled) return;
        this.unsubscribePointerMove();
        this.resetRatingToValue();
    };

    render() {
        const isReadonly = this.props.isReadonly || this.props.isDisabled;

        return (
            <div
                role="slider"
                aria-valuenow={ this.props.value || 0 }
                aria-valuemax={ this.props.to }
                aria-valuemin={ this.props.from }
                tabIndex={ 0 }
                onKeyDown={ (e) => !isReadonly && this.onKeyDown(e) }
                className={ cx(css.container, this.props.isDisabled && uuiMod.disabled, isReadonly && css.containerReadonly, this.props.cx) }
                onPointerMove={ (e) => !isReadonly && this.onPointerMove(e) }
                onPointerLeave={ () => !isReadonly && this.onPointerLeave() }
                onPointerUp={ (e) => !isReadonly && this.onPointerUp(e) }
                onTouchEnd={ (e) => !isReadonly && this.onTouchEnd(e) }
                ref={ this.handleRef }
                { ...this.props.rawProps }
            >
                {this.props.renderRating(this.state.rating, this.getMarkWidth(), this.getNumberOfMarks())}
            </div>
        );
    }
}
