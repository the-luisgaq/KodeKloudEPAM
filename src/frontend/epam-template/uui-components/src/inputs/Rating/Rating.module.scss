.root {
    display: flex;
}

.container {
    &:focus {
        outline: none;
    }

    :global(.uui-rating-empty-star-container) {
        display: flex;
    }

    :global(.uui-rating-filled-star-container) {
        display: flex;
        overflow: hidden;
        position: absolute;
        top: 0;
        left: 0;
    }

    :global(.uui-rating-star) {
        flex-shrink: 0;
    }
}


