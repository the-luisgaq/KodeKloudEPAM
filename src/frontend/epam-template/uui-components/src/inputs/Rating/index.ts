export * from './BaseRating';
export * from './Rating';
