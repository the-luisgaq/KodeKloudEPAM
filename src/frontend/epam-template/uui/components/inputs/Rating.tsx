import { Icon, Overwrite, withMods } from '@epam/uui-core';
import * as uuiComponents from '@epam/uui-components';
import { Tooltip } from '../overlays/Tooltip';
import { settings } from '../../settings';

import css from './Rating.module.scss';

interface RatingMods {
    /**
     *  Rating icon can be changed according to your needs.
     *  Icon can be a React element (usually an SVG element).
     */
    icon?: Icon;
    /**
     * Defines component size.
     * @default '18'
     */
    size?: 18 | 24 | 30;
}
export interface RatingModsOverride {}

/** Represents the 'Core properties' for the Rating component. */
export type RatingCoreProps = Omit<uuiComponents.RatingProps, 'filledStarIcon' | 'emptyStarIcon' | 'renderRating' | 'from' | 'to'>;

/** Represents the properties for the Rating component. */
export interface RatingProps extends RatingCoreProps, Overwrite<RatingMods, RatingModsOverride> {}

function applyRatingMods(mods: RatingProps) {
    return [
        css.root,
        `uui-size-${mods.size || settings.rating.sizes.default}`,
    ];
}

export const Rating = withMods<uuiComponents.RatingProps, RatingProps>(
    uuiComponents.Rating,
    applyRatingMods,
    (props) => ({
        filledStarIcon: props.icon || settings.rating.icons.filledRatingIcon,
        emptyStarIcon: props.icon || settings.rating.icons.emptyRatingIcon,
        Tooltip,
    }),
);
