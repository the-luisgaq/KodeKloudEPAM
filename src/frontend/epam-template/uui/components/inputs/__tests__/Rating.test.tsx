import React from 'react';
import { Rating } from '../Rating';
import { renderSnapshotWithContextAsync } from '@epam/uui-test-utils';
import { ReactComponent as FavoriteIcon } from '@epam/assets/icons/communication-favorite-fill.svg';

describe('Rating', () => {
    it('should be rendered correctly', async () => {
        const tree = await renderSnapshotWithContextAsync(<Rating value={ null } onValueChange={ jest.fn } />);

        expect(tree).toMatchSnapshot();
    });

    it('should be rendered correctly with more props', async () => {
        const tree = await renderSnapshotWithContextAsync(<Rating value={ null } onValueChange={ jest.fn } size={ 24 } step={ 1 } />);

        expect(tree).toMatchSnapshot();
    });

    it('should be rendered with other icon', async () => {
        const tree = await renderSnapshotWithContextAsync(<Rating icon={ FavoriteIcon } value={ null } onValueChange={ jest.fn } size={ 24 } step={ 1 } />);

        expect(tree).toMatchSnapshot();
    });
});
