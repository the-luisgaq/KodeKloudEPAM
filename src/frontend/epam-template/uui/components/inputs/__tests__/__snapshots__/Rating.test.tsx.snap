// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rating should be rendered correctly 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div
      aria-valuemax="5"
      aria-valuemin="1"
      aria-valuenow="0"
      class="container root uui-size-18"
      role="slider"
      tabindex="0"
    >
      <div
        class="container"
      >
        <div
          class="uui-rating-empty-star-container"
        >
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
        </div>
        <div
          class="uui-rating-filled-star-container"
          style="width: 0px;"
        >
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Rating should be rendered correctly with more props 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div
      aria-valuemax="5"
      aria-valuemin="1"
      aria-valuenow="0"
      class="container root uui-size-24"
      role="slider"
      tabindex="0"
    >
      <div
        class="container"
      >
        <div
          class="uui-rating-empty-star-container"
        >
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
        </div>
        <div
          class="uui-rating-filled-star-container"
          style="width: 0px;"
        >
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Rating should be rendered with other icon 1`] = `
<DocumentFragment>
  <div
    class="root"
  >
    <div
      aria-valuemax="5"
      aria-valuemin="1"
      aria-valuenow="0"
      class="container root uui-size-24"
      role="slider"
      tabindex="0"
    >
      <div
        class="container"
      >
        <div
          class="uui-rating-empty-star-container"
        >
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
        </div>
        <div
          class="uui-rating-filled-star-container"
          style="width: 0px;"
        >
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
          <div
            class="uui-rating-star"
          >
            <div
              class="container uui-icon uui-enabled"
            >
              <svg
                class=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
